### PR TITLE
Gitmanager and JWT changes

### LINKED ISSUE #768: Import course (aplus-json) sometimes breaks foreign key constraints when modifying existing courses
Import a course to A+. Then, modify chapter and exercise keys, e.g., by renaming chapter RST files in the course repo or moving chapters to other modules. Exercise keys can also be changed in the chapter RST files in the RST submit directive parameters. After modifications, build the course and import it into A+ again. (Do not empty the database so that the old course structure exists in A+ before the second import.) Does A+ show errors in the import and does it finish successfully?

It may make a difference if there are submissions in the exercises before running the second import because A+ will not delete exercises that have submissions. If the old exercise key is not  part of the new course (aplus-json), then A+ changes the exercise status to "hidden". What if the parent content chapter has been deleted?

**Teacher's original request:**

I got the following error while trying to update my course (Edit course - import). However, pushing the button (Import) again solved the problem.

> Palvelin palautti virheen (server returned error): 'insert or update on table
> "exercise_learningobject" violates foreign key constraint
> "D5881233ddf2ad4d825c0b6aef345c53" DETAIL: Key (parent_id)=(29984) is
> not present in table "exercise_learningobject". ’

**More background:**

The foreign key from learningobject to learningobject typically
refers to exercises that are children of a chapter. If you move
chapters between modules or rename chapter RST files, then
A+ creates new objects with new keys and deletes the objects
with the old unused keys. Then there are these rare cases when
something seems to be deleted too soon or the system somehow
breaks the foreign key constraints.

### PR DESCRIPTION
**Pull request by lainets**
# Description

**What?**

Adds JWT authentication and authorization, and changes related to the new gitmanager.

**Why?**

Gitmanager has adds many errors checks and reliability changes, which require small changes on A+ to work correctly. JWT authorization is needed for authentication between services. This allows e.g. putting course files behind a authorization wall in gitmanager. Also adds automatic course updates, and fixes some configure_url errors for that.

**How?**

Grader authentication is done through JWTs. Course changes are published at the end of the database transaction to make sure that the changes are OK, and the changes happen at roughly the same time.

Fixes #768

# Testing

**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [ ] Django unit tests.
- [ ] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [X] Manual testing.

Automatic tests yet to be added. Does not change the UI, so accesibility tests are not relevant.

Tested manually:
- exercises work
- automatic publishing works
- emailing course instance contacts work

**Did you test the changes in**

- [X] Chrome
- [ ] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [x] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

Error messages for configuring and JWT login.

# Programming style

- [X] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?

# Have you updated the README or other relevant documentation?

- [X] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

### CODE REVIEW

File: edit_course/operations/configure.py [hunk 1]
```diff
@@ -356,299 +357,357 @@ def get_build_log(instance):
...
-            if dt:
-                module.opening_time = dt
-        if not module.opening_time:
-            module.opening_time = instance.starting_time
+        return False, errors
 
-        if "close" in m:
-            dt = parse_date(m["close"], errors)
+    # wrap everything in a transaction to make sure invalid configuration isn't saved
+    with transaction.atomic():
```
Comment (Mankro) [inline edit_course/operations/configure.py:405]: The diff includes everything since all lines were indented. Did you change anything besides the "publish_url" in the end?

File: edit_course/operations/configure.py [hunk 2]
```diff
@@ -356,299 +357,357 @@ def get_build_log(instance):
...
-                module.delete()
+                    module.status = CourseModule.STATUS.HIDDEN
+                    module.save()
+
+        # Clean up obsolete categories.
+        for category in instance.categories.filter(status=LearningObjectCategory.STATUS.HIDDEN):
+            if category.learning_objects.count() == 0:
+                category.delete()
+
+        if "publish_url" in config:
```
Comment (Mankro) [inline edit_course/operations/configure.py:665]: What is this new field "publish_url"? I don't know its purpose.

File: edit_course/operations/configure.py [hunk 3]
```diff
@@ -356,299 +357,357 @@ def get_build_log(instance):
...
+                            _("PUBLISH_ERROR_JSON_PARSER_FAILED -- {e}, {text}"),
+                            e=e,
+                            text=response.text
+                        )]
+                    else:
+                        if isinstance(publish_errors, dict):
+                            success = publish_errors.get("success", True)
+                            publish_errors = publish_errors.get("errors", [])
+
+                        if isinstance(publish_errors, list):
```
Comment (Mankro) [inline edit_course/operations/configure.py:693]:
```suggestion
                        elif isinstance(publish_errors, list):
```

File: edit_course/operations/configure.py [hunk 4]
```diff
@@ -356,299 +357,357 @@ def get_build_log(instance):
...
+        # Clean up obsolete categories.
+        for category in instance.categories.filter(status=LearningObjectCategory.STATUS.HIDDEN):
+            if category.learning_objects.count() == 0:
+                category.delete()
+
+        if "publish_url" in config:
+            success = False
+            publish_errors = []
+            try:
+                response = aplus_auth.get(config["publish_url"])
```
Comment (Mankro) [inline edit_course/operations/configure.py:669]: Where is this HTTP request typically going to? Git manager? And why is this needed?

**Review by Mankro — CHANGES_REQUESTED** [review]
You have done a great job! :+1: 

I want to emphasize that in A-plus, `User.id != UserProfile.id`. This is especially true in our production database since UserProfiles did not exist in the first version of the software. At least in `exercise/permissions.py`, I think you have made a mistake with this and it must be fixed.

Like in the mooc-grader, I think error handling should be done in a bit different way. I have commented on those lines.

Commit messages should start with a capital letter, but let's not spend time on changing the history.


I must admit I don't understand most of the JWT token code. I don't have previous experience with these tokens. I trust you know what you are doing. :smile: 

Do the unit tests and Selenium tests fail? Is there any severe issue behind the failures?

File: exercise/api/views.py [hunk 5]
```diff
@@ -105,45 +102,37 @@ def grader_detail(self, request, *args, **kwargs):
...
-        if not student_id and student_id != 0:
+        user_id = info.get("user_id")
+        if not user_id and user_id != 0:
             raise PermissionDenied(
                 "There is no user_id stored in your grader authentication token, "
                 "so it can't be used to create new submission."
             )
         try:
-            student = UserProfile.objects.get(user_id=student_id)
+            student = UserProfile.objects.get(user_id=user_id)
```
[the remark this answers is not in the document]
Comment (lainets) [reply exercise/api/views.py:120]: It was UserProfile before so it should be now too. Why would it change?

File: edit_course/operations/configure.py (hunk 1 above)
Comment (lainets) [reply edit_course/operations/configure.py:405]: I added rollbacks to the error points "transaction.set_rollback(True)"

Comment (Mankro) [reply edit_course/operations/configure.py:405]: That's good!

File: authorization/api/views.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    GET assumes the payload is sent as a part of a JWT token.
+
+    POST allows you to specify the payload fields in the POST parameters.
+    """
+    permission_classes = [IsAuthenticated]
+
+    def get_audience(self, alias: str) -> str:
+        return url_to_audience(alias)
+
+    def get_expiration_time(self, request: Request, payload: Payload): #TODO: return None for no expiration
```
[the remark this answers is not in the document]
Comment (Mankro) [reply authorization/api/views.py]: I agree that all tokens should expire at some point.

File: edit_course/operations/configure.py (hunk 2 above)
Comment (lainets) [reply edit_course/operations/configure.py:665]: It is used to notify gitmanager (or whatever service we are fetching the config from) that the configuration seems OK and now we should publish the changes

File: edit_course/operations/configure.py (hunk 3 above)
Comment (lainets) [reply edit_course/operations/configure.py:693]: This is on purpose: `publish_errors = publish_errors.get("errors", [])` above might set it to a list

File: edit_course/operations/configure.py (hunk 4 above)
Comment (lainets) [reply edit_course/operations/configure.py:669]: Like I responded to another comment: It is used to notify gitmanager (or whatever service we are fetching the config from) that the configuration seems OK and now we should publish the changes. The point is that gitmanager does not publish (roll out) the changes before we know that A+ is ok with them, and it also means that A+ does its preprocessing _before_ any changes are published, so downtime (or time when the course is invalid, in some sense) is short.

File: exercise/api/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Mankro) [reply exercise/api/views.py:120]: Yup. Actually, when I look at this again, then obviously the UserProfile is fetched with a User.id using the foreign key relation. All good!

**Comment by lainets** [comment]
> Do the unit tests and Selenium tests fail? Is there any severe issue behind the failures?
A single unit test is failing due to a missing setting. I'll be fixing it tomorrow

**Comment by Mankro** [comment]
I merged this now so that we can deploy it to the testing server. More changes and fixes may be done afterwards.

File: doc/DEPLOYMENT.md [hunk 7]
```diff
@@ -67,6 +67,12 @@ Common system configuration
           | sudo tee /etc/tmpfiles.d/aplus.conf > /dev/null
         sudo systemd-tmpfiles --create
 
+ 1. Create RSA keys for [JWT authentication](AUTH.md)
+
+        # generate private key
+        openssl genrsa -out private.pem 2048
```
[the remark this answers is not in the document]
Comment (lainets) [reply doc/DEPLOYMENT.md:73]: No source I found says that 2048 bit keys are outdated

File: doc/GRADERS.md [hunk 8]
```diff
@@ -4,6 +4,15 @@ A+ grader service protocols
 The philosophy of A+ is to integrate external grader services into the course environment.
 The graders should be stateless and need only to worry about grading a submission.
 
+## Authentication and Authorization
+
+Every incoming and outgoing request must either have a A+-provided grading
+token as a GET parameter (included in the `submission_url` below) or a JWT
```
[the remark this answers is not in the document]
Comment (lainets) [reply doc/GRADERS.md:10]: The A+ protocol gives the token as part of the submission_url. This would require changing the protocol. Other than that, the code change is trivial. But, if we do this change, we probably should just require the use of the JWT token instead,

File: exercise/api/views.py [hunk 9]
```diff
@@ -87,7 +84,7 @@ class ExerciseViewSet(mixins.RetrieveModelMixin,
         url_path='grader',
         url_name='grader',
         methods=['get', 'post'],
-        permission_classes = GRADER_PERMISSION,
+        get_permissions = lambda: [JWTSubmissionCreatePermission()],
```
[the remark this answers is not in the document]
Comment (lainets) [reply exercise/api/views.py:87]: The permissions system is stupid because of the inheritance: if you want to access an exercise , you also need to access a course instance. The grader only has access to that exercise and not the instance, so it doesn't work.

File: exercise/permissions.py [hunk 10]
```diff
@@ -16,7 +18,7 @@
 )
 
 
-class ExerciseVisiblePermission(ObjectVisibleBasePermission):
+class ExerciseVisiblePermissionBase(ObjectVisibleBasePermission):
```
[the remark this answers is not in the document]
Comment (lainets) [reply exercise/permissions.py:22]: Putting the Base after makes it easier to search and is easier to see at a glance, which is why I think this way is better. The meaning is also technically a little different here: ExerciseVisiblePermissionBase is not a base class for ExerciseVisiblePermission but a part of it

File: locale/en/LC_MESSAGES/django.po [hunk 11]
```diff
@@ -1866,6 +1911,24 @@ msgstr "Category requires a name."
...
+#, fuzzy, python-brace-format
+#| msgid "PUBLISH_RESPONSE_NON_200 -- {status_code}"
+msgid "PUBLISH_RESPONSE_NON_200 -- {status_code}"
+msgstr "Publishing responded with status {response.status_code}"
+
+#: edit_course/operations/configure.py
+#, python-brace-format
+msgid "PUBLISH_ERROR_JSON_PARSER_FAILED -- {e}, {text}"
+msgstr ""
+"Failed to load publish response JSON: {e}:\n"
```
[the remark this answers is not in the document]
Comment (lainets) [reply locale/en/LC_MESSAGES/django.po:1924]: "publish" refers to the endpoint that is used to publish (the changes to) a course. I couldn't quickly come up with a better message, so I put that. Maybe "publishing" instead of "publish"?

File: doc/AUTH.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
+# Authorization and Authentication protocol
+
+Every request A+ makes to graders and other services contain a JWT
+authentication/authorization token. Any request to A+ that requires
+authentication must also contain a JWT. This does not include submission
+grading, as then it includes a token in the GET parameters. Though, this token
+can be sent as part of the JWT's tokens field.
```
[the remark this answers is not in the document]
Comment (lainets) [reply doc/AUTH.md:7]: Better now?

**Review by lainets — COMMENTED** [review]
Fixes in PR #933

File: doc/GRADERS.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (Mankro) [reply doc/GRADERS.md:10]: Yes indeed, it wasn't first clear to me that the GET query parameters refer to the old grader protocol whereas the HTTP Authorization header refers to the new JWT tokens.

File: locale/en/LC_MESSAGES/django.po (hunk 11 above)
[the remark this answers is not in the document]
Comment (Mankro) [reply locale/en/LC_MESSAGES/django.po:1924]: "Failed to load the publishing response JSON:"
